### PR TITLE
warnings to logging and changing defaults for write_map, read_map

### PR DESCRIPTION
**Pull request by zonca**
Following #675 and previous issues, I switched the library to using `logging`, most of the previous warnings became debug or info messages which are not shown by default, just a minor important ones became `logging.warning`.

The most important objective of warnings was to warn users of the future change of the default setting for `hp.write_map` and `hp.read_map`.
Until the last release (1.14.0), whatever precision a map is in memory, it is written to disk in single precision (`float32`) and when read, it is converted to double precision (`float64`). As was written in the warning, now instead `healpy` will maintain whatever precision the map is in both when reading and writing.
I am a bit worried that a user upgrading is not notified about this, it is in the docstring and the changelog but there is no other notification, any suggestions here?

The other warning that was present was in `synalm` and `synfast` about the order of  `C_ell`, going for `new=False` to `new=True`, basically changing from TT TE EE.... to TT EE BB. This seems too dangerous for old code. I prefer to remove this notice and leave the default as it is.

### CODE REVIEW

File: healpy/fitsfunc.py [hunk 1]
```diff
@@ -190,14 +184,8 @@ def write_map(
...
-        warnings.warn(
-            "The default dtype of write_map() will change in a future version: "
-            "explicitly set the dtype if it is important to you",
-            category=FutureWarning,
-        )
-        dtype = [np.float32 for x in m]
-        # Change this at some point to:
-        # dtype = [x.dtype for x in m]
+        dtype = [x.dtype for x in m]
+        logging.info("setting the output map dtype to %s", str(dtype))
```
Comment (lpsinger) [inline healpy/fitsfunc.py:188]:
```suggestion
        logging.info("setting the output map dtype to %s", dtype)
```

File: healpy/fitsfunc.py [hunk 2]
```diff
@@ -135,7 +124,12 @@ def write_map(
     extra_header=(),
     overwrite=False,
 ):
-    """Writes a healpix map into a healpix file.
+    """Writes a healpix map into a healpix FITS file.
+
+    WARNING: starting from healpy 1.15.0, if you do not specify `dtype`,
+    the map will be written to disk with the same precision it is stored in memory.
+    Previously, by default `healpy` wrote maps in `float32`.
+    To reproduce the same behaviour of `healpy` 1.14.0 and below, set `dtype=np.float32`.
```
Comment (lpsinger) [inline healpy/fitsfunc.py:132]:
```suggestion
    .. warning:: Starting from healpy 1.15.0, if you do not specify `dtype`,
       the map will be written to disk with the same precision it is stored in memory.
       Previously, by default `healpy` wrote maps in `float32`.
       To reproduce the same behaviour of `healpy` 1.14.0 and below, set `dtype=np.float32`.
```

Also change similar text from other docstrings to ReStructuredText admonitions so that they are given emphasis.

File: healpy/__init__.py [hunk 3]
```diff
@@ -107,16 +105,9 @@
...
-    ``verbose=False``.
-    Warnings can be re-enabled calling ``hp.enable_warnings()``.
-    """
-    warnings.filterwarnings(action="ignore", module="healpy")
-
-
-def enable_warnings():
-    warnings.simplefilter("always")
\ No newline at end of file
+logging.getLogger("healpy").addHandler(logging.NullHandler())
```
Comment (lpsinger) [inline healpy/__init__.py:113]: It would be a good idea to see if there is any benefit to directly tapping into Astropy's warning system.

File: healpy/fitsfunc.py [hunk 4]
```diff
@@ -347,9 +338,8 @@ def read_map(
       the header number to look at (start at 0)
     h : bool, optional
       If True, return also the header. Default: False.
-    verbose : bool, optional
-      If True, print a number of diagnostic messages, call hp.disable_warnings() to
-      disable warnings for all functions.
+    verbose : bool, not supported anymore
```
Comment (lpsinger) [inline healpy/fitsfunc.py:341]: Deprecate this so that it can be removed in a future release. Astropy's utils module has some features for deprecating arguments.

File: healpy/sphtfunc.py [hunk 5]
```diff
@@ -1242,8 +1200,7 @@ def beam2bl(beam, theta, lmax):
 
     nx = len(theta)
     nb = len(beam)
-    if nb != nx:
-        warnings.warn("beam and theta must have same size!")
+    assert nb == nx, "beam and theta must have same size!"
```
Comment (lpsinger) [inline healpy/sphtfunc.py:1203]: Seems like a ValueError would be more appropriate here than an assertion.

File: healpy/test/test_visufunc.py [hunk 6]
```diff
@@ -93,7 +93,7 @@ def test_azeqview_ma_nocrash(self):
     def test_cmap_colors(self):
         # Get a built-in colormap
         name = 'viridis'
-        cmap = matplotlib.cm.get_cmap(name)
+        cmap = copy.copy(matplotlib.cm.get_cmap(name))
```
Comment (lpsinger) [inline healpy/test/test_visufunc.py:96]: Not sure why this is part of this changeset.

**Comment by zonca** [comment]
@kadrlica would you please review this PR? suggestions very welcome

**Review by lpsinger — CHANGES_REQUESTED** [review]
Instead of using the global logging helper functions, please get a logger object and use it.

```python
log = logging.getLogger("healpy")
...
log.warning(...)
```

File: healpy/projaxes.py [hunk 7]
```diff
@@ -915,7 +891,7 @@ def create_colormap(cmap, badcolor="gray", bgcolor="white"):
             "newcm", cmap0._segmentdata, cmap0.N
         )
     else:
-        newcm = copy.deepcopy(cmap0)
+        newcm = copy.copy(cmap0)
```
Comment (zonca) [inline healpy/projaxes.py:894]: @kadrlica compared to https://github.com/healpy/healpy/pull/627, I was getting a warning with `deepcopy` and not with `copy`

Comment (lpsinger) [reply healpy/projaxes.py:894]: So this should probably go in a separate PR.

Comment (zonca) [reply healpy/projaxes.py]: right, moved to https://github.com/healpy/healpy/pull/681

File: healpy/fitsfunc.py (hunk 2 above)
Comment (zonca) [reply healpy/fitsfunc.py]: see ca0473f863a3d299e0e9b4d07fffa3e8df547218

File: healpy/fitsfunc.py (hunk 4 above)
Comment (zonca) [reply healpy/fitsfunc.py:342]: this would generate warnings. isn't it acceptable to ask users to go through their code and remove all `verbose` arguments? it seems a quick fix that can be just done once (and will need to be done in the future).

Comment (lpsinger) [reply healpy/fitsfunc.py:342]: A deprecation warning is least obtrusive way to tell users to do that.

Comment (lpsinger) [reply healpy/fitsfunc.py:342]: https://docs.astropy.org/en/stable/api/astropy.utils.decorators.deprecated_renamed_argument.html#astropy.utils.decorators.deprecated_renamed_argument

Comment (zonca) [reply healpy/fitsfunc.py]: ok

File: doc/index.rst [hunk 8]
```diff
@@ -23,6 +23,33 @@ is very welcome to propose new features.
...
+
+    import warnings
+    from astropy.utils.exceptions import AstropyDeprecationWarning
+    warnings.simplefilter('ignore', category=AstropyDeprecationWarning)
+
+Changelog
+---------
+
+Review the changes in each release in the `CHANGELOG on Github <https://github.com/healpy/healpy/blob/main/CHANGELOG.rst>`_.
+
```
Comment (zonca) [inline doc/index.rst:52]: @lpsinger ok, I deprecated all verbose, added docs here about verbosity

File: healpy/__init__.py [hunk 9]
```diff
@@ -107,16 +107,54 @@
...
+    log.setLevel(level)
+
+    # create console handler
+    handler = logging.StreamHandler()
+
+    # create formatter
+    formatter = logging.Formatter(log_format)
+    handler.setFormatter(formatter)
+
+    log.addHandler(handler)
```
Comment (zonca) [inline healpy/__init__.py:160]: @lpsinger it is not straightforward to configure `logging`, setting only the level is not enough you also need to add a handler, so better provide a function.

File: healpy/__init__.py (hunk 3 above)
Comment (zonca) [reply healpy/__init__.py]: I think `logging` only except for deprecation is fine.

File: doc/index.rst [hunk 10]
```diff
@@ -23,6 +23,33 @@ is very welcome to propose new features.
...
+For more details see the `Python documentation <https://docs.python.org/3/library/logging.html>`_.
+
+All the `verbose` keywords (except :py:func:remove_dipole) are now deprecated and
+will be removed in the future.
+You can disable deprecation warnings with::
+
+    import warnings
+    from astropy.utils.exceptions import AstropyDeprecationWarning
+    warnings.simplefilter('ignore', category=AstropyDeprecationWarning)
```
Comment (zonca) [inline doc/index.rst:46]: @lpsinger ok, I deprecated all verbose, added docs here about verbosity

File: healpy/__init__.py (hunk 9 above)
Comment (zonca) [inline healpy/__init__.py:160]: @lpsinger it is not straightforward to configure logging, setting only the level is not enough you also need to add a handler, so better provide a function.

File: healpy/pixelfunc.py [hunk 11]
```diff
@@ -1578,7 +1574,7 @@ def remove_dipole(
 
         lon, lat = R.vec2dir(dipole, lonlat=True)
         amp = np.sqrt((dipole * dipole).sum())
-        warnings.warn(
+        print(
```
Comment (kadrlica) [inline healpy/pixelfunc.py:1577]: Should this also be a logging statement?

File: healpy/projaxes.py [hunk 12]
```diff
@@ -46,9 +47,8 @@ class SphericalProjAxes(matplotlib.axes.Axes):
     coordprec : number of digit after floating point for coordinates display.
     format : format string for value display.
 
-    Notes
-    -----
-    Other keywords from Axes (see Axes).
+    .. note::
+       Other keywords from Axes (see Axes).
```
Comment (lpsinger) [inline healpy/projaxes.py:51]: More incidental format changes that should go in a separate PR or changeset.

File: healpy/projaxes.py [hunk 13]
```diff
@@ -194,8 +189,9 @@ def projmap(
         if vmin == vmax:
             vmin -= 1.0
             vmax += 1.0
-        cm, nn = get_color_table(vmin, vmax, img[w], cmap=cmap, norm=norm,
-                                 badcolor=badcolor, bgcolor=bgcolor)
+        cm, nn = get_color_table(
+            vmin, vmax, img[w], cmap=cmap, norm=norm, badcolor=badcolor, bgcolor=bgcolor
+        )
```
Comment (lpsinger) [inline healpy/projaxes.py:194]: Incidental code style change, should go in a separate changeset.

File: healpy/projaxes.py [hunk 14]
```diff
@@ -205,8 +201,6 @@ def projmap(
             norm=nn,
             interpolation="nearest",
             origin="lower",
-            vmin=vmin,
-            vmax=vmax,
```
Comment (lpsinger) [inline healpy/projaxes.py:209]: Another unrelated change...

File: healpy/projaxes.py [hunk 15]
```diff
@@ -293,9 +286,11 @@ def projplot(self, *args, **kwds):
...
                     linestyle, marker, color = matplotlib.axes._process_plot_format(fmt)
                 except:  # matplotlib 1.4 and later
-                    linestyle, marker, color = matplotlib.axes._axes._process_plot_format(
-                        fmt
-                    )
+                    (
+                        linestyle,
+                        marker,
+                        color,
+                    ) = matplotlib.axes._axes._process_plot_format(fmt)
```
Comment (lpsinger) [inline healpy/projaxes.py:293]: Another unrelated change...

File: healpy/projaxes.py [hunk 16]
```diff
@@ -536,16 +527,8 @@ def graticule(
...
-            warnings.warn(
-                "{0} {1} {2} {3}".format(
-                    pmin / dtor, pmax / dtor, mmin / dtor, mmax / dtor
-                )
-            )
         if not kwds.pop("force", False):
-            dpar, dmer = self._get_interv_graticule(
-                pmin, pmax, dpar, mmin, mmax, dmer, verbose=verbose
-            )
+            dpar, dmer = self._get_interv_graticule(pmin, pmax, dpar, mmin, mmax, dmer)
```
Comment (lpsinger) [inline healpy/projaxes.py:531]: Another unrelated change...

File: healpy/projaxes.py [hunk 17]
```diff
@@ -612,19 +595,18 @@ def graticule(
...
+        """Delete all graticules previously created on the Axes."""
         if hasattr(self, "_graticules"):
             for dum1, dum2, g in self._graticules:
                 for gl in g:
                     for l in gl:
                         if l in self.lines:
                             self.lines.remove(l)
                         else:
-                            warnings.warn("line not in lines")
+                            log.warning("line not in lines")
```
Comment (lpsinger) [inline healpy/projaxes.py:606]: Is this going to spew a large amount of log chatter, since it's inside a triply nested loop?

File: healpy/rotator.py [hunk 18]
```diff
@@ -55,18 +58,6 @@
     .value
 )
 
-
-class ConsistencyWarning(Warning):
-    """Warns for a problem in the consistency of data
-    """
-
-    pass
```
Comment (lpsinger) [inline healpy/rotator.py:63]: Needs a deprecation warning prior to removal.

**Comment by zonca** [comment]
splitting into multiple PRs
